File: apps/api/src/worldcup_api/schemas/predictions.py
from datetime import datetime

from pydantic import BaseModel, Field


class ScoringRulesSchema(BaseModel):
    correct_result: int = Field(default=2, ge=0)
    correct_goal_difference: int = Field(default=3, ge=0)
    exact_score: int = Field(default=4, ge=0)


class TeamSide(BaseModel):
    id: int | None = None
    name: str


class MatchSummary(BaseModel):
    id: int
    source: str | None = None
    source_match_id: str | None = None
    date: datetime
    stage: str
    group_name: str | None = None
    home_team: TeamSide
    away_team: TeamSide
    venue: str | None = None
    status: str = "scheduled"


class ScoreProbability(BaseModel):
    score: str
    p: float


class RecommendedTip(BaseModel):
    score: str
    expected_points: float
    explanation: str


class TeamRating(BaseModel):
    team: str
    model_elo: int
    strength_score: float
    tier: str
    known_rating: bool


class PredictionConfidence(BaseModel):
    label: str
    score: float
    reason: str


class ModelContext(BaseModel):
    model_version: str
    data_source: str
    training_status: str
    rating_source: str
    explanation: list[str]


class MatchPreviewResponse(BaseModel):
    fixture: str
    tactical_preview: str
    key_factors: list[str]
    upset_scenario: str
    injury_watch: list[str]
    source_urls: list[str]
    confidence: float


class PredictionResponse(BaseModel):
    match: MatchSummary
    p_home_win: float
    p_draw: float
    p_away_win: float
    lambda_home: float
    lambda_away: float
    most_likely_scores: list[ScoreProbability]
    recommended_tip: RecommendedTip
    home_rating: TeamRating
    away_rating: TeamRating
    rating_delta: int
    confidence: PredictionConfidence
    model_context: ModelContext
    model_notes: list[str]


class MatchListResponse(BaseModel):
    matches: list[MatchSummary]


class PredictionListResponse(BaseModel):
    predictions: list[PredictionResponse]

File: apps/web/lib/predictions.ts
export type ApiMatch = {
  id: number;
  source: string | null;
  source_match_id: string | null;
  date: string;
  stage: string;
  group_name: string | null;
  home_team: { id: number | null; name: string };
  away_team: { id: number | null; name: string };
  venue: string | null;
  status: string;
  home_score: number | null;
  away_score: number | null;
};

export type TipCandidate = {
  score: string;
  expected_points: number;
  exact_probability: number;
  label: string;
  rationale: string;
  actual_points: number | null;
};

export type PredictionResponse = {
  match: ApiMatch;
  p_home_win: number;
  p_draw: number;
  p_away_win: number;
  lambda_home: number;
  lambda_away: number;
  most_likely_scores: { score: string; p: number }[];
  recommended_tip: {
    score: string;
    expected_points: number;
    explanation: string;
    actual_points: number | null;
    actual_score: string | null;
  };
  tip_candidates: TipCandidate[];
  home_rating: { model_elo: number; strength_score: number; tier: string; known_rating: boolean };
  away_rating: { model_elo: number; strength_score: number; tier: string; known_rating: boolean };
  rating_delta: number;
  confidence: { label: "Low" | "Medium" | "High"; score: number; reason: string };
  model_context: {
    model_version: string;
    data_source: string;
    training_status: string;
    rating_source: string;
    explanation: string[];
  };
  model_notes: string[];
};

export type ScoringRules = {
  exactScore: number;
  goalDiff: number;
  result: number;
};

export function apiBaseUrl() {
  return (process.env.NEXT_PUBLIC_API_BASE_URL ?? "").replace(/\/$/, "");
}

export function formatPercent(value: number) {
  return `${Math.round(value * 100)}%`;
}

export function formatKickoff(value: string) {
  return new Date(value).toLocaleString();
}

export function favoriteLabel(prediction: PredictionResponse) {
  const outcomes = [
    { label: prediction.match.home_team.name, value: prediction.p_home_win },
    { label: "Draw", value: prediction.p_draw },
    { label: prediction.match.away_team.name, value: prediction.p_away_win }
  ];
  return outcomes.sort((a, b) => b.value - a.value)[0];
}

export function conservativePriorPick(prediction: PredictionResponse) {
  const delta = prediction.rating_delta;
  if (Math.abs(delta) < 75) {
    return "1-1";
  }
  return delta > 0 ? "1-0" : "0-1";
}

export function candidateByScore(prediction: PredictionResponse, score: string) {
  return prediction.tip_candidates.find((candidate) => candidate.score === score) ?? null;
}
